File: apps/mobile/modules/remodex-secure-transport/ios/RemodexSecureTransportModule.swift
import ExpoModulesCore

public class RemodexSecureTransportModule: Module {
  public func definition() -> ModuleDefinition {
    Name("RemodexSecureTransport")

    Events("stateChanged", "plaintextMessage", "secureError", "socketClosed")

    Function("getState") { () -> [String: Any?] in
      return [
        "status": "idle",
        "transportMode": "unsupported",
        "connectedMacDeviceId": nil,
        "relay": nil,
        "sessionId": nil,
        "trustedMacs": [],
        "lastError": nil
      ]
    }

    AsyncFunction("listTrustedMacs") { () -> [[String: Any?]] in
      return []
    }

    AsyncFunction("forgetTrustedMac") { (_ macDeviceId: String) -> [String: Any?] in
      return [
        "status": "idle",
        "transportMode": "unsupported",
        "connectedMacDeviceId": nil,
        "relay": nil,
        "sessionId": nil,
        "trustedMacs": [],
        "lastError": nil
      ]
    }

    AsyncFunction("connectFromQr") { (_ payload: [String: Any]) -> [String: Any?] in
      return [
        "status": "error",
        "transportMode": "unsupported",
        "connectedMacDeviceId": payload["macDeviceId"] as? String,
        "relay": payload["relay"] as? String,
        "sessionId": payload["sessionId"] as? String,
        "trustedMacs": [],
        "lastError": "Native secure transport is not implemented yet."
      ]
    }

    AsyncFunction("connectTrusted") { (_ macDeviceId: String) -> [String: Any?] in
      return [
        "status": "error",
        "transportMode": "unsupported",
        "connectedMacDeviceId": macDeviceId,
        "relay": nil,
        "sessionId": nil,
        "trustedMacs": [],
        "lastError": "Native secure transport is not implemented yet."
      ]
    }

    AsyncFunction("disconnect") { () -> [String: Any?] in
      return [
        "status": "idle",
        "transportMode": "unsupported",
        "connectedMacDeviceId": nil,
        "relay": nil,
        "sessionId": nil,
        "trustedMacs": [],
        "lastError": nil
      ]
    }

    AsyncFunction("sendPlaintext") { (_ text: String) in
      throw Exception(name: "ERR_REMODEX_TRANSPORT_UNAVAILABLE", description: "Native secure transport is not implemented yet.")
    }
  }
}

File: apps/mobile/modules/remodex-secure-transport/android/src/main/java/expo/modules/remodexsecuretransport/RemodexSecureTransportModule.kt
package expo.modules.remodexsecuretransport

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class RemodexSecureTransportModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("RemodexSecureTransport")

    Events(
      "stateChanged",
      "plaintextMessage",
      "secureError",
      "socketClosed"
    )

    Function("getState") {
      mapOf(
        "status" to "idle",
        "transportMode" to "unsupported",
        "connectedMacDeviceId" to null,
        "relay" to null,
        "sessionId" to null,
        "lastError" to null,
        "trustedMacs" to emptyList<Map<String, Any?>>()
      )
    }

    AsyncFunction("listTrustedMacs") {
      emptyList<Map<String, Any?>>()
    }

    AsyncFunction("forgetTrustedMac") { _: String ->
      mapOf(
        "status" to "idle",
        "transportMode" to "unsupported",
        "connectedMacDeviceId" to null,
        "relay" to null,
        "sessionId" to null,
        "lastError" to null,
        "trustedMacs" to emptyList<Map<String, Any?>>()
      )
    }

    AsyncFunction("connectFromQr") { payload: Map<String, Any?> ->
      val macDeviceId = payload["macDeviceId"] as? String
      mapOf(
        "status" to "error",
        "transportMode" to "unsupported",
        "connectedMacDeviceId" to macDeviceId,
        "relay" to null,
        "sessionId" to null,
        "lastError" to "Native secure transport is not implemented yet.",
        "trustedMacs" to emptyList<Map<String, Any?>>()
      )
    }

    AsyncFunction("connectTrusted") { macDeviceId: String ->
      mapOf(
        "status" to "error",
        "transportMode" to "unsupported",
        "connectedMacDeviceId" to macDeviceId,
        "relay" to null,
        "sessionId" to null,
        "lastError" to "Native secure transport is not implemented yet.",
        "trustedMacs" to emptyList<Map<String, Any?>>()
      )
    }

    AsyncFunction("disconnect") {
      sendEvent(
        "socketClosed",
        mapOf(
          "code" to 1000,
          "reason" to "Disconnected"
        )
      )
      mapOf(
        "status" to "idle",
        "transportMode" to "unsupported",
        "connectedMacDeviceId" to null,
        "relay" to null,
        "sessionId" to null,
        "lastError" to null,
        "trustedMacs" to emptyList<Map<String, Any?>>()
      )
    }

    AsyncFunction("sendPlaintext") { text: String ->
      sendEvent(
        "plaintextMessage",
        mapOf("text" to text)
      )
      mapOf("ok" to true)
    }
  }
}
